sklearn/impute.py

```diff
@@ -138,30 +169,52 @@ def fit(self, X, y=None):
         -------
         self : SimpleImputer
         """
-        # Check parameters
-        allowed_strategies = ["mean", "median", "most_frequent"]
-        if self.strategy not in allowed_strategies:
-            raise ValueError("Can only use these strategies: {0} "
-                             " got strategy={1}".format(allowed_strategies,
-                                                        self.strategy))
+        X = self._validate_input(X)
 
-        X = check_array(X, accept_sparse='csc', dtype=FLOAT_DTYPES,
-                        force_all_finite='allow-nan'
-                        if self.missing_values == 'NaN'
-                        or np.isnan(self.missing_values) else True)
+        # default missing_values is 0 for numerical input and "missing_value"
+        # otherwise
+        if self.fill_value is None:
+            if X.dtype.kind in ("i", "f"):
+                fill_value = 0
+            else:
+                fill_value = "missing_value"
+        else:
+            fill_value = self.fill_value
+
+        # fill_value should be numerical in case of numerical input
+        if self.strategy == "constant":
+            if X.dtype.kind in ("i", "f"):
+                if not isinstance(fill_value, numbers.Real):
+                    raise TypeError(
+                        "fill_value={0} is invalid. Expected a numerical value"
+                        " to numerical data".format(fill_value))
+
+            elif X.dtype.kind == "O":
+                if not isinstance(fill_value, six.string_types):
+                    raise TypeError(
+                        "fill_value={0} is invalid. Expected an str instance "
+                        "when imputing categorical data.".format(fill_value))
```

Comment:
I am not sure it is good / needed to be that strict here. For example, you can have Categorical data in pandas with integer categories (or other non-string values) where it can make sense to fill with another value (and `column.dtype.kind` will be "O" for categorical columns)

Reply:
We should have a test for this specific use case: imputing a categorical column with a pandas categorical value.

Reply:
maybe 
if not (isinstance(fill_value, six.string_types) or isinstance(fill_value, numbers.Reals) or fill_value is None)
is enough ?

Reply:
We talked IRL with @jorisvandenbossche and if the categorical datafram has integer values with missing values, then the check_array conversion will yield a floating point array so its already covered by the existing tests.

Let's stay strict in this checks for now. If a user reports a valid use case for being laxer we can always change that later.